lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,444 @@
+        self.controllers = list()
+        self.bad_controllers = list()
+        self.validation_list = None
+
+    def add_controller(self, controller):
+        self.controllers.append(controller)
+
+    def add_bad_controller(self, controller):
+        self.bad_controllers.append(controller)
+
+    def add_error_or_inconceivable_entries(self, val, controller, dirs):
+        self.validation_list[val][controller] = dirs
+
+    def add_unexpected_entries(self, val, controller, dirs):
+        """Gather values for each validation_list dict key"""
+        self.validation_list[val][controller] = dirs
+
+    def header(self, fp, status):
+        """Adds main starting and ending messages for each Hierarchy"""
+        lead_newline = "\n" if status == "start" else ""
+        fp.write(
+            f"{lead_newline}\n{status}-{self.config.TS[4:]}: {self.name} hierarchy: {self.path}\n"
+        )
+        return
+
+    def check_controller_in_list(self, controller):
+        """Checks whether there are controllers in dictionary or not"""
+        return any(
+            [
+                controller in self.validation_list[key]._dict
+                for key in self.validation_list
+            ]
+        )
+
+    """
+    ##FIXME: The mention of hierarchies(archive, controller) from here on, in
+            Hierarchy class is mainly to keep the format algned with the shell script
+    """
+
+    def dump_check(self, fp, controller, key, cnt, hierarchy=False):
+        """Output Messages for each list of controllers, files,
+        directories, etc depicting their cause of being added to
+        that list or dictionary
+        """
+        lead_asterisk = "\t* " if hierarchy == "archive" else "\t"
+        if controller in self.validation_list[key]._dict:
+            if key not in ["subdir_status_indicators", "prefix_status_indicators"]:
+                fp.write(f"{lead_asterisk}{self.validation_list[key]._get_msg()}\n")
+                if hierarchy == "archive":
+                    cnt = self.output_format(
+                        fp, self.validation_list[key]._dict[controller], cnt, "archive"
+                    )
+                else:
+                    cnt = self.output_format(
+                        fp, self.validation_list[key]._dict[controller], cnt
+                    )
+            elif "subdirs" in self.validation_list[key]._dict[controller]:
+                fp.write(f"{lead_asterisk}{self.validation_list[key]._get_msg()}\n")
+                cnt += 1
+            elif "prefix_dir" in self.validation_list[key]._dict[controller]:
+                fp.write(f"{lead_asterisk}{self.validation_list[key]._get_msg()}\n")
+                cnt += 1
+        return cnt
+
+    def check_tab_format(self, hierarchy):
+        """Check the format of tab acc. to hierarchy"""
+
+        lead_tab = ""
+        if hierarchy == "controller":
+            lead_tab = "\t"
+        elif hierarchy == "archive":
+            lead_tab = "\t  "
```

Comment:
Differences like these should be implicit in the class hierarchy, not handled by rippling explicit arguments through the call tree. That is, there's no reason that the base `Hierarchy` can't have a `self.lead_tab = ""`, which each subclass overrides in its `__init__` to the appropriate value. Then the methods can just reference `self.lead_tab` instead of repeatedly passing and processing the `hierarchy` parameter.

Reply:
Done